### ISSUE TITLE
Build circle layer driven by Multiple Indicators

### ISSUE THREAD

**Issue by kahummer** [onaio/gisida#433]
At the ward level, the client would like to view the **population affected** and **damage sites** on a single layer.  The current implementation employs [mapbox-gl-js filter expressions](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/)  to control circle-radius based on total population data and circle-color based on infrastructure damage.
![Screenshot from 2020-02-19 14-42-07](https://user-images.githubusercontent.com/16018123/74831317-14243000-5326-11ea-809c-0187eec794f2.png)

We may have to implement code that drives both indicators on the legend @engwiri should we have a legend that shows both or have a legend showing only one indicator?

**Comment by engwiri** [comment]
@kahummer this looks good. A few comments:
- The legend only needs to be driven by the colour, I don't think we need both but I may be wrong
- Can we use the same colours in color brewer so we can see the larger concentration of damaged sites with darker colours
- Is it possible in this map and the expressions to use the filters to choose between sectors so that the colours change based on the variable you are using? 
- What will be in the pop up?

**Comment by kahummer** [comment]
Sure we can use color brewer for this. This is what we have on the popup currently
![Screenshot from 2020-02-19 15-34-41](https://user-images.githubusercontent.com/16018123/74835340-2609d100-532e-11ea-9b89-4e3e562ce1e8.png). 
Am not certain we have expressions tied to our filters. The current one's we have used are being applied at layer spec level. @cKellyDesign @KipSigei please share your thoughts on this when you have a moment

**Comment by kahummer** [comment]
Colorbrewer added ![Screenshot from 2020-02-19 17-00-57](https://user-images.githubusercontent.com/16018123/74845103-8bb28900-533f-11ea-8ace-263603e3fca5.png)

**Comment by cKellyDesign** [comment]
@engwiri - It's not very clear to me [from the data](https://github.com/onaio/idai-recovery-zimbabwe/blob/add-fill-ward/public/data/cleanedcombinedzim.csv) how this layer is supposed to be visualizing the total number of damaged sites.

Is there a column in [this CSV](https://github.com/onaio/idai-recovery-zimbabwe/blob/add-fill-ward/public/data/cleanedcombinedzim.csv) that is a sum total number of damaged sites?

If not then we would have to aggregate multiple columns together (in order for it to be filterable b/c we would have to re-run the aggregation with filtered data), which is not something Gisida currently does, it only aggregates across rows of a single column.
Is there a column in [this CSV](https://github.com/onaio/idai-recovery-zimbabwe/blob/add-fill-ward/public/data/cleanedcombinedzim.csv) that is a sum total number of damaged sites?

If not then we would have to aggregate multiple columns together (in order for it to be filterable b/c we would have to re-run the aggregation with filtered data), which is not something Gisida currently does, it only aggregates across rows of a single column.

**Comment by cKellyDesign** [comment]
> It's not very clear to me from the data how this layer is supposed to be visualizing the total number of damaged sites.

The key word here being `how`, because it is not clear from the column headers:
1. which single column represents the total number of damaged sites? if we don't have one,
2. what are all the columns are to be summed together (in a new column)?

We need this single column because:

> [If we were to programmatically calculate the number of damaged sites from multiple columns] we would have to aggregate multiple columns together, which is not something Gisida currently does, it only aggregates across rows of a single column.

Which means the answer to the following request is "No we cannot without developing new features": 

> Is it possible in this map and the expressions to use the filters to choose between sectors so that the colours change based on the variable you are using?

**Comment by cKellyDesign** [comment]
@engwiri - Should the number of sites per ward be weighted? Meaning should 1-damaged-site-out-of-2-sites-in-the-ward be visualized differently than 1-damaged-site-out-of-50-sites-in-the-ward?

**Comment by cKellyDesign** [comment]
> What will be in the pop up?

This needs to be defined by @engwiri or @akahure

**Comment by engwiri** [comment]
Let me try and define this a little better. The need that the client has expressed was a clear way to identify and compare areas that have more structural damage after the cyclone but weighted against the population. So for example, if ward a has 5 out of 10 damaged infrastructure points against a population of 2500 and ward b has 2 out of 3 damaged sites against a population of 10,000; there should be a way to visually represent that ward b has more needs than a. 

To do this, we wanted to first see if it was possible to visually represent both parameters on the site, so circle layers where the radius is driven by population size and colour ranges representing damage intensity seemed to be a good fit. But on further reflection with Matt, this still doesn't communicate the right message although it would be a nice quick win. Ultimately what I think we need is:

1. Determine a ratio of damaged to not damaged for each sector and across sectors as separate variables - @Janemesi 
2. Show circle fills with radius driven by population (as we have done) and the fill colours are driven by the ratio of damaged to not damaged sites.
3. The pop up should show: Number of damaged sites, Total number of sites, ratio*, partners working in that area and lastly a narrative showing the relationship of damaged sites to the population, for example, _2/5 schools are functional in a ward with 10,000 people. So each school is serving 5,000 people instead of 2,000. UNOPS and WFP are working in this area._

Lastly, it makes sense that we do this across sectors so that as the initial layer, the ideal would be to have them all in one layer and then have filters to select sectors ie. water, sanitation, education etc and see the same relationships either for each individual sector or when you combine water and sanitation. I realise this may require a different data model given whats in gisida so if @kahummer, @cKellyDesign, @KipSigei can suggest the format that @Janemesi should prepare. Please add that to the issue as a comment below.

**Comment by engwiri** [comment]
@kahummer the colour brewer codes you added are showing ward population rather than number of sites. Can we correct that?

>Is there a column in this CSV that is a sum total number of damaged sites?
@cKellyDesign yes the _inf_damaged_ column is the total number of sites and it covers more than just the 3 damaged sites there, there are about 10 categories and they are all in the CSV. 

Please review these requirements before we go back to the client with anything. This is not an easy client, I want to make sure we exploring everything before we can push back

**Comment by kahummer** [comment]
@engwiri The circle-color is driven by inf_damaged the misleading part is the text on the legend I changed that to Damaged sites per ward

**Comment by cKellyDesign** [comment]
> Let me try and define this a little better.

@engwiri Success! 🙇 This makes sense to me and feels like an effective and efficient way to show that `Ward B` needs more help than `Ward A` in your scenario.

> yes the `inf_damaged` column is the total number of sites and it covers more than just the 3 damaged sites there

@engwiri ah I see makes sense, I didn't realize there were other columns being in added in 👍 

> Determine a ratio of damaged to not damaged for each sector and across sectors as separate variables

@engwiri @Janemesi - If we were to take that ratio and multiply it against `population`, could/would that number be an accurate representation of `effected population` at the ward level? If so, what if we used that number for both circle size and color so as to "double encode" the visualization? In this scenario, bigger circles would always be redder.

> the ideal would be to have them all in one layer and then have filters to select sectors

@engwiri I agree that this would be the ideal, such a layer with such filtering would have to be able to aggregate the totals / weighted indicators in the browser, which Gisida isn't currently capable of doing because it cannot aggregate across columns. Even if it were, however, I don't know how we could implement the type of weighted indicator / effected population calculations described above. I suggest we:
- [ ] Add the initial layer for all damaged sites across sectors
- [ ] Add layers for sector specific sites

**Comment by kahummer** [comment]
@cKellyDesign @ngwiri We had a discussion with @KipSigei  and we think it would be possible to have a filter simmillar to what we have on nutrition sites, where all wards will have a column which has all sectors with destroyed sites. @Janemesi has added thi column. 
somnutrition.onalabs.org

**Comment by cKellyDesign** [comment]
@kahama94 - I'm not sure how this would work, aren't those filter options derived from site level data? With a single, static, weighted indicator from a CSV being used to drive the visualization (circle color and/or size), **we have no way to programmatically re-calculate the filtered results**.

**Comment by kahummer** [comment]
@ckelly In the somnutrition sites we have a column that holds site services provided at every nutrition site in a multi select manner eg otp, sc, bsfp etc In our case we would build the filter based on a column with water points, sanitation etc on every row based on presence of damage site.
 When we apply a filter e.g water points that will filter out all wards with no damaged water points. I agree The approach of programmatically re-calculating the filtered results is not achievable with this type of data that's why we though the multi select filters would be an achievable fix

**Comment by cKellyDesign** [comment]
> When we apply a filter e.g water points that will filter out all wards with no damaged water points.

Interesting, I guess that will technically work, but I think it gets away from answering the question: `which ward needs most help, considering damaged points that are not water points?`

Instead your approach answers the question: `which wards do not have damaged water sites?` (which is different in my option)

@engwiri what are your thoughts?

**Comment by kahummer** [comment]
@engwiri  @cKellyDesign an update on current progress. I'm thinking the values/ratios on the popup and legend  should be in percentages
![Screenshot from 2020-02-21 16-06-47](https://user-images.githubusercontent.com/16018123/75037179-9cd8d280-54c4-11ea-8984-6aadcf37b959.png)
For the filters 
![Screenshot from 2020-02-21 16-08-11](https://user-images.githubusercontent.com/16018123/75037208-ac581b80-54c4-11ea-81b4-d5b8bfa6b71b.png)

**Comment by cKellyDesign** [comment]
@kahummer - what's the status of this issue? Has it been approved by @engwiri, @akahure, and/or the client?

**Comment by engwiri** [comment]
@cKellyDesign this is exactly what we were looking for so @kahummer please confirm that this is done.

**Comment by kahummer** [comment]
@cKellyDesign @engwiri The status of this issue is it's **in review**. Has been deployed http://idairecovery-zimbabwe.onalabs.org/

**Comment by kahummer** [comment]
@cKellyDesign @KipSigei  Can we drive the popup based on the filter datasets. An example below
![Screenshot from 2020-03-05 15-55-17](https://user-images.githubusercontent.com/16018123/75983607-e0d2cb00-5ef9-11ea-988f-c28297802777.png)
Is it possible to show the number of damaged sites for water points only on the popup and remove everything unrelated to water points on the popup ?

**Comment by cKellyDesign** [comment]
> Is it possible to show the number of damaged sites for water points only on the popup and remove everything unrelated to water points on the popup ?

@kahummer I don't believe so, I think currently using [Mustache templates](https://mustache.github.io/mustache.5.html) we can't really introduce that sort of conditional logic (using Mustache syntax).

**Comment by kahummer** [comment]
cc @engwiri https://github.com/onaio/gisida/issues/433#issuecomment-595307490

**Comment by ciremusyoka** [comment]
> @cKellyDesign @KipSigei Can we drive the popup based on the filter datasets. An example below
> ![Screenshot from 2020-03-05 15-55-17](https://user-images.githubusercontent.com/16018123/75983607-e0d2cb00-5ef9-11ea-988f-c28297802777.png)
> Is it possible to show the number of damaged sites for water points only on the popup and remove everything unrelated to water points on the popup ?

@kahummer @cKellyDesign My thought on this is we have another `template` on client project other than the `popup template` (maybe `filterPopup`) which we can append the selected filter with its value only.
But with this the `Number of damaged sites` will not be entirely accurate because we will be showing only one damaged site data and the `Number of damaged sites` is a summation of all damaged sites. For example, assuming we have `2 damaged sites` and we select `water points` filter, but in that ward we have `0 damaged water points`.  Will this not raise concerns?

**Comment by cKellyDesign** [comment]
> @kahummer @cKellyDesign My thought on this is we have another template on client project other than the popup template (maybe filterPopup) which we can append the selected filter with its value only.

I'm not sure how this would be accomplished with our [`addPopup` code](https://github.com/onaio/gisida/blob/master/src/map/addPopUp.js), also isn't the `template` defined in the layer spec? 

I don't know that we should be committing to developing new popup features here (at least this sounds like a new feature), but I'm curious as to what @engwiri thinks of the popup.

**Comment by ciremusyoka** [comment]
@kahummer I might have been wrong considering the data is dynamic and if we were to use my approach it will mean defining a `template` for each filter which is not possible.

**Comment by cKellyDesign** [comment]
@engwiri is this in review with the client?

**Comment by cKellyDesign** [comment]
@engwiri @kahummer @ciremusyoka - What's the status of this deliverable? Can this ticket be closed?

**Comment by engwiri** [comment]
@akahure did the client ever review this and provide feedback. We need to know if this is what they wanted and whether to keep it

**Comment by akahure** [comment]
@engwiri I did show it to them and they asked that we should leave it as it is. 
when I asked you about this, you mentioned since we cannot achieve the specific requirement we should put it on hold, for now, I .e.the requirements on the filter for the popup. 

I think we can close this @cKellyDesign
